File: regions/system/library/typescript/bb-ai-client/__tests__/bb-ai-client.test.ts
import { vi, describe, it, expect } from 'vitest';
import {
  AiClientError,
  InMemoryAiClient,
  HttpAiClient,
  type CompleteRequest,
  type EmbedRequest,
} from '../src/index.js';

// InMemoryAiClient のテスト
describe('InMemoryAiClient', () => {
  it('デフォルトでモックレスポンスを返すこと', async () => {
    const client = new InMemoryAiClient();
    const res = await client.complete({
      model: 'test-model',
      messages: [{ role: 'user', content: 'hello' }],
    });
    expect(res.id).toBe('mock-id');
    expect(res.model).toBe('test-model');
    expect(res.content).toBe('mock response');
  });

  it('カスタム complete 関数を使用できること', async () => {
    const client = new InMemoryAiClient({
      complete: async (req) => ({
        id: 'custom-id',
        model: req.model,
        content: 'custom response',
        usage: { inputTokens: 10, outputTokens: 20 },
      }),
    });
    const res = await client.complete({
      model: 'gpt-4',
      messages: [{ role: 'user', content: 'test' }],
    });
    expect(res.id).toBe('custom-id');
    expect(res.usage.inputTokens).toBe(10);
  });

  it('embed でデフォルト空埋め込みを返すこと', async () => {
    const client = new InMemoryAiClient();
    const res = await client.embed({ model: 'embed-model', texts: ['hello', 'world'] });
    expect(res.model).toBe('embed-model');
    expect(res.embeddings).toHaveLength(2);
  });

  it('listModels でデフォルト空配列を返すこと', async () => {
    const client = new InMemoryAiClient();
    const models = await client.listModels();
    expect(models).toEqual([]);
  });
});

// HttpAiClient のテスト
describe('HttpAiClient', () => {
  it('complete で正しいエンドポイントにPOSTすること', async () => {
    const mockFetch = vi.fn().mockResolvedValue({
      ok: true,
      json: async () => ({
        id: 'resp-id',
        model: 'claude-3',
        content: 'Hello!',
        usage: { input_tokens: 5, output_tokens: 10 },
      }),
    });
    vi.stubGlobal('fetch', mockFetch);

    const client = new HttpAiClient({ baseUrl: 'https://api.example.com', apiKey: 'test-key' });
    const res = await client.complete({
      model: 'claude-3',
      messages: [{ role: 'user', content: 'Hi' }],
    });

    expect(mockFetch).toHaveBeenCalledWith(
      'https://api.example.com/v1/complete',
      expect.objectContaining({ method: 'POST' }),
    );
    expect(res.content).toBe('Hello!');
    expect(res.usage.inputTokens).toBe(5);
    expect(res.usage.outputTokens).toBe(10);

    vi.unstubAllGlobals();
  });

  it('embed で正しいエンドポイントにPOSTすること', async () => {
    const mockFetch = vi.fn().mockResolvedValue({
      ok: true,
      json: async () => ({
        model: 'embed-v1',
        embeddings: [[0.1, 0.2], [0.3, 0.4]],
      }),
    });
    vi.stubGlobal('fetch', mockFetch);

    const client = new HttpAiClient({ baseUrl: 'https://api.example.com' });
    const res = await client.embed({ model: 'embed-v1', texts: ['a', 'b'] });

    expect(mockFetch).toHaveBeenCalledWith(
      'https://api.example.com/v1/embed',
      expect.objectContaining({ method: 'POST' }),
    );
    expect(res.embeddings).toHaveLength(2);

    vi.unstubAllGlobals();
  });

  it('listModels で正しいエンドポイントにGETすること', async () => {
    const mockFetch = vi.fn().mockResolvedValue({
      ok: true,
      json: async () => [
        { id: 'model-1', name: 'Model One', description: 'desc' },
      ],
    });
    vi.stubGlobal('fetch', mockFetch);

    const client = new HttpAiClient({ baseUrl: 'https://api.example.com' });
    const models = await client.listModels();

    expect(mockFetch).toHaveBeenCalledWith(
      'https://api.example.com/v1/models',
      expect.objectContaining({ method: 'GET' }),
    );
    expect(models).toHaveLength(1);
    expect(models[0].id).toBe('model-1');

    vi.unstubAllGlobals();
  });

  it('APIエラー時に AiClientError をスローすること', async () => {
    const mockFetch = vi.fn().mockResolvedValue({
      ok: false,
      status: 401,
      statusText: 'Unauthorized',
    });
    vi.stubGlobal('fetch', mockFetch);

    const client = new HttpAiClient({ baseUrl: 'https://api.example.com' });
    await expect(
      client.complete({ model: 'test', messages: [{ role: 'user', content: 'hi' }] }),
    ).rejects.toThrow(AiClientError);

    vi.unstubAllGlobals();
  });
});

// AiClientError のテスト
describe('AiClientError', () => {
  it('正しいフィールドを持つこと', () => {
    const err = new AiClientError('test error', 500);
    expect(err.message).toBe('test error');
    expect(err.statusCode).toBe(500);
    expect(err.name).toBe('AiClientError');
    expect(err).toBeInstanceOf(Error);
  });
});

File: regions/system/library/dart/bb-ai-client/test/bb_ai_client_test.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

import 'package:k1s0_bb_ai_client/bb_ai_client.dart';

void main() {
  // InMemoryAiClient のテスト
  group('InMemoryAiClient', () {
    late InMemoryAiClient client;

    setUp(() {
      client = InMemoryAiClient();
    });

    test('デフォルトでモックレスポンスを返すこと', () async {
      final res = await client.complete(CompleteRequest(
        model: 'test-model',
        messages: [const ChatMessage(role: 'user', content: 'hello')],
      ));
      expect(res.id, equals('mock-id'));
      expect(res.model, equals('test-model'));
      expect(res.content, equals('mock response'));
    });

    test('カスタム complete 関数を使用できること', () async {
      final customClient = InMemoryAiClient(
        complete: (req) async => CompleteResponse(
          id: 'custom-id',
          model: req.model,
          content: 'custom',
          usage: const Usage(inputTokens: 10, outputTokens: 20),
        ),
      );
      final res = await customClient.complete(CompleteRequest(
        model: 'gpt-4',
        messages: [const ChatMessage(role: 'user', content: 'test')],
      ));
      expect(res.id, equals('custom-id'));
      expect(res.usage.inputTokens, equals(10));
    });

    test('embed でデフォルト空埋め込みを返すこと', () async {
      final res = await client.embed(EmbedRequest(
        model: 'embed-model',
        texts: ['hello', 'world'],
      ));
      expect(res.model, equals('embed-model'));
      expect(res.embeddings, hasLength(2));
    });

    test('listModels でデフォルト空配列を返すこと', () async {
      final models = await client.listModels();
      expect(models, isEmpty);
    });
  });

  // HttpAiClient のテスト
  group('HttpAiClient', () {
    test('complete で正しいエンドポイントにPOSTすること', () async {
      final mockClient = MockClient((request) async {
        expect(request.url.path, equals('/v1/complete'));
        expect(request.method, equals('POST'));
        return http.Response(
          json.encode({
            'id': 'resp-id',
            'model': 'claude-3',
            'content': 'Hello!',
            'usage': {'input_tokens': 5, 'output_tokens': 10},
          }),
          200,
        );
      });

      final client = HttpAiClient(
        baseUrl: 'https://api.example.com',
        apiKey: 'test-key',
        httpClient: mockClient,
      );
      final res = await client.complete(CompleteRequest(
        model: 'claude-3',
        messages: [const ChatMessage(role: 'user', content: 'Hi')],
      ));

      expect(res.content, equals('Hello!'));
      expect(res.usage.inputTokens, equals(5));
      expect(res.usage.outputTokens, equals(10));
    });

    test('embed で正しいエンドポイントにPOSTすること', () async {
      final mockClient = MockClient((request) async {
        expect(request.url.path, equals('/v1/embed'));
        return http.Response(
          json.encode({
            'model': 'embed-v1',
            'embeddings': [[0.1, 0.2], [0.3, 0.4]],
          }),
          200,
        );
      });

      final client = HttpAiClient(
        baseUrl: 'https://api.example.com',
        httpClient: mockClient,
      );
      final res = await client.embed(EmbedRequest(
        model: 'embed-v1',
        texts: ['a', 'b'],
      ));
      expect(res.embeddings, hasLength(2));
    });

    test('listModels で正しいエンドポイントにGETすること', () async {
      final mockClient = MockClient((request) async {
        expect(request.url.path, equals('/v1/models'));
        expect(request.method, equals('GET'));
        return http.Response(
          json.encode([
            {'id': 'model-1', 'name': 'Model One', 'description': 'desc'},
          ]),
          200,
        );
      });

      final client = HttpAiClient(
        baseUrl: 'https://api.example.com',
        httpClient: mockClient,
      );
      final models = await client.listModels();
      expect(models, hasLength(1));
      expect(models.first.id, equals('model-1'));
    });

    test('APIエラー時に AiClientError をスローすること', () async {
      final mockClient = MockClient((_) async => http.Response('Unauthorized', 401));

      final client = HttpAiClient(
        baseUrl: 'https://api.example.com',
        httpClient: mockClient,
      );
      expect(
        () => client.complete(CompleteRequest(
          model: 'test',
          messages: [const ChatMessage(role: 'user', content: 'hi')],
        )),
        throwsA(isA<AiClientError>()),
      );
    });
  });

  // AiClientError のテスト
  group('AiClientError', () {
    test('正しいフィールドを持つこと', () {
      const err = AiClientError('test error', statusCode: 500);
      expect(err.message, equals('test error'));
      expect(err.statusCode, equals(500));
      expect(err.toString(), contains('500'));
    });
  });
}
